In [87]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook. 
# If you get an error here, make sure that mysql and pymysql are installed correctly. 

%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [88]:
# Establish a connection to the local database using the '%sql' magic command.
# Replace 'password' with our connection password and `db_name` with our database name. 
# If you get an error here, please make sure the database name or password is correct.

%sql mysql+pymysql://root:password@localhost:3306/md_water_services_two

'Connected: root@md_water_services_two'

In [89]:
%%sql

# To show all the tables in the database
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
8 rows affected.


[('data_dictionary',),
 ('employee',),
 ('global_water_access',),
 ('location',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',)]

In [90]:
%%sql


SELECT * FROM md_water_services.data_dictionary;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
49 rows affected.


[('employee', 'assigned_employee_id', 'Unique ID assigned to each employee', 'INT', 'visits'),
 ('employee', 'employee_name', 'Name of the employee', 'VARCHAR(255)', ''),
 ('employee', 'phone_number', 'Contact number of the employee', 'VARCHAR(15)', ''),
 ('employee', 'email', 'Email address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'address', 'Residential address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'town_name', 'Name of the town where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'province_name', 'Name of the province where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'position', 'Position or job title of the employee', 'VARCHAR(255)', ''),
 ('visits', 'record_id', 'Unique ID assigned to each visit', 'int', 'water_quality, water_source'),
 ('visits', 'location_id', 'ID of the location visited', 'varchar(255)', 'location'),
 ('visits', 'source_id', 'ID of the water source visited', 'varchar(510)', 'well_pollution'),
 ('visits', 'time_of_record', 'Date and time of the visit', 'datetime', ''),
 ('visits', 'visit_count', 'Number of visits made to this location', 'int', ''),
 ('visits', 'time_in_queue', 'Time spent by people waiting for water in a queue at the location', 'int', ''),
 ('visits', 'assigned_employee_id', 'ID of the employee who visited the location', 'int', 'employee'),
 ('water_quality', 'record_id', 'Unique ID assigned to each record', 'int', ''),
 ('water_quality', 'subjective_quality_score', 'Score representing the subjective quality of the water source', 'int', ''),
 ('water_quality', 'visit_count', 'Number of visits made for data collection', 'int', ''),
 ('water_source', 'source_id', 'Unique ID assigned to each water source', 'VARCHAR(510)', 'visits'),
 ('water_source', 'type_of_water_source', 'Type or category of the water source. Can be : tap_in_home, tap_in_home_broken, well, shared_tap, river', 'VARCHAR(255)', ''),
 ('water_source', 'number_of_people_served', 'Number of people served by this water source', 'INT', ''),
 ('well_pollution', 'source_id', 'ID of the water source being tested for pollution', 'varchar(258)', 'visits'),
 ('well_pollution', 'date', 'Date of the pollution test', 'datetime', ''),
 ('well_pollution', 'description', 'Description of the pollution test', 'varchar(255)', ''),
 ('well_pollution', 'pollutant_ppm', 'Result of the pollution test in parts per million', 'float', ''),
 ('well_pollution', 'biological', 'Biological contamination level', 'float', ''),
 ('well_pollution', 'results', "Result of the pollution test. Can be 'dirty', 'biologically contaminated', or 'clean'", 'varchar(255)', ''),
 ('location', 'location_id', 'Unique ID assigned to each location', 'varchar(255)', 'visits'),
 ('location', 'address', 'Address of the location', 'varchar(255)', ''),
 ('location', 'province_name', 'Name of the province where the location is situated', 'varchar(255)', ''),
 ('location', 'town_name', 'Name of the town where the location is situated', 'varchar(255)', ''),
 ('location', 'location_type', 'Type or category of the location', 'varchar(255)', ''),
 ('global_water_access', 'name', 'The country or area name', 'varchar(255)', ''),
 ('global_water_access', 'region', 'Geographical region', 'varchar(255)', ''),
 ('global_water_access', 'year', 'Year of the data record', 'int', ''),
 ('global_water_access', 'pop_n', 'The national population size estimate in thousands', 'float', ''),
 ('global_water_access', 'pop_u', 'The urban population share estimate in percentage points (%)', 'float', ''),
 ('global_water_access', 'wat_bas_n', 'The estimated national share of people with at least basic service (%)', 'float', ''),
 ('global_water_access', 'wat_lim_n', 'The estimated national share of people with limited service (%)', 'float', ''),
 ('global_water_access', 'wat_unimp_n', 'The estimated national share of people with unimproved service (%)', 'float', ''),
 ('global_water_access', 'wat_sur_n', 'The estimated national share of people with surfa

In [91]:
%%sql

SELECT * FROM md_water_services.location LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('AkHa00000', '2 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00001', '10 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00002', '9 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00003', '139 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00004', '17 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban')]

In [92]:
%%sql

SELECT * FROM md_water_services.visits LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

In [93]:
%%sql

SELECT * FROM md_water_services.water_source LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942)]

In [94]:
%%sql

SELECT * FROM md_water_services.water_quality LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1), (3, 10, 1), (4, 4, 1)]

### Employee table. It has info on all of our workers, but note that the email addresses have not been added. We will have to send them reports and figures, so let's update it. Luckily the emails for our department are easy: first_name.last_name@ndogowater.gov.

### Method 1

In [95]:
%%sql


SELECT * FROM md_water_services.employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [96]:
%%sql

SELECT
    employee_name,
    LENGTH(employee_name) AS string_length,
    POSITION(' ' IN employee_name) AS post_first_name,
    LEFT(employee_name, POSITION(' ' IN employee_name) -1) AS first_name,
    LENGTH(LEFT(employee_name, POSITION(' ' IN employee_name) -1)) AS first_name_len
FROM 
    md_water_services.employee LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Amara Jengo', 11, 6, 'Amara', 5),
 ('Bello Azibo', 11, 6, 'Bello', 5),
 ('Bakari Iniko', 12, 7, 'Bakari', 6),
 ('Malachi Mavuso', 14, 8, 'Malachi', 7),
 ('Cheche Buhle', 12, 7, 'Cheche', 6),
 ('Zuriel Matembo', 14, 7, 'Zuriel', 6),
 ('Deka Osumare', 12, 5, 'Deka', 4),
 ('Lalitha Kaburi', 14, 8, 'Lalitha', 7),
 ('Enitan Zuri', 11, 7, 'Enitan', 6),
 ('Farai Nia', 9, 6, 'Farai', 5)]

In [97]:
%%sql

SELECT
    employee_name,
    LENGTH(employee_name) AS string_length,

    SUBSTRING(
        employee_name
        FROM POSITION(' ' IN employee_name) + 1
    ) AS last_name,

    LENGTH(
        SUBSTRING(
            employee_name
            FROM POSITION(' ' IN employee_name) + 1
        )
    ) AS last_name_len

FROM md_water_services.employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Amara Jengo', 11, 'Jengo', 5),
 ('Bello Azibo', 11, 'Azibo', 5),
 ('Bakari Iniko', 12, 'Iniko', 5),
 ('Malachi Mavuso', 14, 'Mavuso', 6),
 ('Cheche Buhle', 12, 'Buhle', 5),
 ('Zuriel Matembo', 14, 'Matembo', 7),
 ('Deka Osumare', 12, 'Osumare', 7),
 ('Lalitha Kaburi', 14, 'Kaburi', 6),
 ('Enitan Zuri', 11, 'Zuri', 4),
 ('Farai Nia', 9, 'Nia', 3)]

In [98]:
%%sql

SELECT
    employee_name,
    LOWER(CONCAT(LEFT(employee_name, POSITION(' ' IN employee_name) -1), '.',  SUBSTRING(employee_name FROM POSITION(' ' IN employee_name) + 1), '@ndogowater.gov.')) AS email
FROM 
    md_water_services.employee
LIMIT 10;
    

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Amara Jengo', 'amara.jengo@ndogowater.gov.'),
 ('Bello Azibo', 'bello.azibo@ndogowater.gov.'),
 ('Bakari Iniko', 'bakari.iniko@ndogowater.gov.'),
 ('Malachi Mavuso', 'malachi.mavuso@ndogowater.gov.'),
 ('Cheche Buhle', 'cheche.buhle@ndogowater.gov.'),
 ('Zuriel Matembo', 'zuriel.matembo@ndogowater.gov.'),
 ('Deka Osumare', 'deka.osumare@ndogowater.gov.'),
 ('Lalitha Kaburi', 'lalitha.kaburi@ndogowater.gov.'),
 ('Enitan Zuri', 'enitan.zuri@ndogowater.gov.'),
 ('Farai Nia', 'farai.nia@ndogowater.gov.')]

### Method 2

In [99]:
%%sql

SELECT
    employee_name,
    REPLACE (employee_name, ' ', '.') AS email
FROM  md_water_services.employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Amara Jengo', 'Amara.Jengo'),
 ('Bello Azibo', 'Bello.Azibo'),
 ('Bakari Iniko', 'Bakari.Iniko'),
 ('Malachi Mavuso', 'Malachi.Mavuso'),
 ('Cheche Buhle', 'Cheche.Buhle'),
 ('Zuriel Matembo', 'Zuriel.Matembo'),
 ('Deka Osumare', 'Deka.Osumare'),
 ('Lalitha Kaburi', 'Lalitha.Kaburi'),
 ('Enitan Zuri', 'Enitan.Zuri'),
 ('Farai Nia', 'Farai.Nia')]

In [100]:
%%sql

SELECT
    employee_name,
    LOWER(CONCAT(REPLACE (employee_name, ' ', '.'), '@ndogowater.gov.')) AS email
FROM  md_water_services.employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Amara Jengo', 'amara.jengo@ndogowater.gov.'),
 ('Bello Azibo', 'bello.azibo@ndogowater.gov.'),
 ('Bakari Iniko', 'bakari.iniko@ndogowater.gov.'),
 ('Malachi Mavuso', 'malachi.mavuso@ndogowater.gov.'),
 ('Cheche Buhle', 'cheche.buhle@ndogowater.gov.'),
 ('Zuriel Matembo', 'zuriel.matembo@ndogowater.gov.'),
 ('Deka Osumare', 'deka.osumare@ndogowater.gov.'),
 ('Lalitha Kaburi', 'lalitha.kaburi@ndogowater.gov.'),
 ('Enitan Zuri', 'enitan.zuri@ndogowater.gov.'),
 ('Farai Nia', 'farai.nia@ndogowater.gov.')]

## UPDATING THE EMAIL COLUMN

In [101]:
%%sql
UPDATE md_water_services.employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov');

SELECT 'update complete' AS status;


 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
56 rows affected.
1 rows affected.


[('update complete',)]

In [102]:
%%sql
SELECT * 
FROM md_water_services.employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (5, 'Zuriel Matembo', '+99034075111', 'zuriel.matembo@ndogowater.gov', '26 Bahari Ya Faraja Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (6, 'Deka Osumare', '+99379364631', 'deka.osumare@ndogowater.gov', '104 Kenyatta Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (7, 'Lalitha Kaburi', '+99681623240', 'lalitha.kaburi@ndogowater.gov', '145 Sungura Amanpour Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (8, 'Enitan Zuri', '+99248509202', 'enitan.zuri@ndogowater.gov', '117 Kampala Road', 'Hawassa', 'Zanzibar', 'Field Surveyor'),
 (10, 'Farai Nia', '+99570082739', 'farai.nia@ndogowater.gov', '33 Angélique Kidjo Avenue', 'Amanzi', 'Dahabu', 'Field Surveyor')]

## The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. However, the phone number has 13 characters, indicating there's an extra character

In [103]:
%%sql



SELECT
    LENGTH(TRIM(phone_number))
FROM
    employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[(12,), (12,), (12,), (12,), (12,), (12,), (12,), (12,), (12,), (12,)]

## Using the employee table to count how many of our employees live in each town.

In [104]:
%%sql
SELECT 
    town_name,
    COUNT(*) AS num_employees
FROM 
    md_water_services.employee
GROUP BY town_name;


 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
9 rows affected.


[('Ilanga', 3),
 ('Rural', 29),
 ('Lusaka', 4),
 ('Zanzibar', 4),
 ('Dahabu', 6),
 ('Kintampo', 1),
 ('Harare', 5),
 ('Yaounde', 1),
 ('Serowe', 3)]

In [105]:
%%sql
SELECT *
FROM visits
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14),
 (5, 'KiRu29315', 'KiRu29315224', datetime.datetime(2021, 1, 1, 10, 17), 1, 9, 40),
 (6, 'AkRu05234', 'AkRu05234224', datetime.datetime(2021, 1, 1, 10, 18), 1, 0, 30),
 (7, 'KiRu28520', 'KiRu28520224', datetime.datetime(2021, 1, 1, 10, 28), 1, 0, 34),
 (8, 'HaZa21742', 'HaZa21742224', datetime.datetime(2021, 1, 1, 10, 37), 1, 0, 6),
 (9, 'AmDa12214', 'AmDa12214224', datetime.datetime(2021, 1, 1, 10, 58), 1, 0, 36)]

## So let's gather some data on their performance in this process, so we know those who really put all their effort in.

In [106]:
%%sql
SELECT
    assigned_employee_id,
    SUM(visit_count) AS number_of_visits
FROM 
    visits
GROUP BY 
    assigned_employee_id
LIMIT
     3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
3 rows affected.


[(0, Decimal('2779')), (1, Decimal('8944')), (2, Decimal('4805'))]

## Understand where the water sources are in Maji Ndogo.

In [107]:
%%sql

SELECT
    *
FROM 
    location
LIMIT
    6;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
6 rows affected.


[('AkHa00000', '2 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00001', '10 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00002', '9 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00003', '139 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00004', '17 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00005', '125 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban')]

In [108]:
%%sql


# Query that counts the number of records per town

SELECT
    COUNT(location_id) AS records_per_town,
    town_name
FROM 
    location
GROUP BY 
    town_name
ORDER BY records_per_town DESC
LIMIT 8;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
8 rows affected.


[(23740, 'Rural'),
 (1650, 'Harare'),
 (1090, 'Amina'),
 (1070, 'Lusaka'),
 (990, 'Mrembo'),
 (930, 'Asmara'),
 (930, 'Dahabu'),
 (780, 'Kintampo')]

In [109]:
%%sql


# Counting the records per province.

SELECT
    COUNT(location_id) AS records_per_province,
    province_name
FROM 
    location
GROUP BY 
    province_name
ORDER BY records_per_province DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[(9510, 'Kilimani'),
 (8940, 'Akatsi'),
 (8220, 'Sokoto'),
 (6950, 'Amanzi'),
 (6030, 'Hawassa')]

In [110]:
%%sql


# Counting the records per town and per province.

SELECT
    province_name,
    town_name,
    COUNT(location_id) AS records_per_town
FROM 
    location
GROUP BY 
    province_name, town_name
ORDER BY 
    province_name, records_per_town DESC
LIMIT
    10

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Akatsi', 'Lusaka', 1070),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Amanzi', 'Rural', 3100),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Amanzi', 'Amina', 670),
 ('Amanzi', 'Pwani', 520),
 ('Amanzi', 'Abidjan', 400)]

In [111]:
%%sql


# Counting the records per location.

SELECT
    COUNT(location_id) AS num_sources,
    location_type
FROM 
    location
GROUP BY location_type;

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
2 rows affected.


[(15910, 'Urban'), (23740, 'Rural')]

## Looking at the water source table

In [112]:
%%sql


# Counting the records per location.

SELECT
    *
FROM 
    water_source
LIMIT 10

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
10 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942),
 ('AkHa00005224', 'tap_in_home', 736),
 ('AkHa00006224', 'tap_in_home', 882),
 ('AkHa00007224', 'tap_in_home', 554),
 ('AkHa00008224', 'well', 398),
 ('AkHa00009224', 'well', 346)]

#### Total number of people we surveyed

In [113]:
%%sql

SELECT
    SUM(number_of_people_served) AS Total_persons_surveyed
FROM 
    water_source

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
1 rows affected.


[(Decimal('27628140'),)]

#### Total number of well, taps and rivers

In [114]:
%%sql

SELECT
    type_of_water_source,
    COUNT(number_of_people_served) AS number_of_sources
FROM 
    water_source
GROUP BY type_of_water_source

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('well', 17383),
 ('shared_tap', 5767),
 ('river', 3379)]

#### Average number of people that are served by each water source

In [115]:
%%sql

SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served), 0) AS ave_people_per_source
FROM 
    water_source
GROUP BY type_of_water_source

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('tap_in_home', Decimal('644')),
 ('tap_in_home_broken', Decimal('649')),
 ('well', Decimal('279')),
 ('shared_tap', Decimal('2071')),
 ('river', Decimal('699'))]

#### The total number of people served by each type of water source in total.

In [116]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS population_served
FROM 
    water_source
GROUP BY type_of_water_source
ORDER BY population_served DESC

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('shared_tap', Decimal('11945272')),
 ('well', Decimal('4841724')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('river', Decimal('2362544'))]

#### The total percentage of people served by each type of water source in total.

In [117]:
%%sql

SELECT
    type_of_water_source,
    ROUND((SUM(number_of_people_served)/27000000)*100, 0) AS percentage_people_per_source
FROM 
    water_source
GROUP BY type_of_water_source
ORDER BY percentage_people_per_source DESC

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
5 rows affected.


[('shared_tap', Decimal('44')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

## Start of a solution
### At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision
### My simple approach is to fix the things that affect most people first.

In [118]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS population_served,
    RANK() OVER( ORDER BY (SUM(number_of_people_served ) )DESC) AS rank_by_population
FROM 
    water_source 
WHERE type_of_water_source != 'tap_in_home'
GROUP BY type_of_water_source

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
4 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home_broken', Decimal('3799720'), 3),
 ('river', Decimal('2362544'), 4)]

In [119]:
%%sql

SELECT
    *,
    RANK() OVER( ORDER BY type_of_water_source, number_of_people_served) AS priority_rank
FROM
water_source
WHERE type_of_water_source != 'tap_in_home'
ORDER BY priority_rank 
LIMIT 6

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
6 rows affected.


[('SoRu38500224', 'river', 400, 1),
 ('SoRu39436224', 'river', 400, 1),
 ('SoRu38759224', 'river', 400, 1),
 ('SoRu38345224', 'river', 400, 1),
 ('SoRu37948224', 'river', 400, 1),
 ('SoRu37633224', 'river', 400, 1)]

### These are some of the things I think are worth looking at:
#### 1. How long did the survey take?

In [120]:
%%sql

SELECT
    *
FROM 
    visits
LIMIT
    6

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
6 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14),
 (5, 'KiRu29315', 'KiRu29315224', datetime.datetime(2021, 1, 1, 10, 17), 1, 9, 40)]

In [121]:
%%sql

SELECT
    DATEDIFF( MAX(time_of_record), MIN(time_of_record)) AS date_difference
FROM 
    visits


 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
1 rows affected.


[(924,)]

#### 2. What is the average total queue time for water?

In [122]:
%%sql

SELECT
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS average_total_queue_time
FROM
visits

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
1 rows affected.


[(Decimal('123'),)]

In [123]:
%%sql

SELECT
    DAYNAME(time_of_record) AS day_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS average_total_queue_time
FROM
visits
GROUP BY day_of_week

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
7 rows affected.


[('Friday', Decimal('120')),
 ('Saturday', Decimal('246')),
 ('Sunday', Decimal('82')),
 ('Monday', Decimal('137')),
 ('Tuesday', Decimal('108')),
 ('Wednesday', Decimal('97')),
 ('Thursday', Decimal('105'))]

In [124]:
%%sql

SELECT
TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
ROUND(AVG(NULLIF(time_in_queue, 0))) AS average_total_queue_time
FROM
visits
GROUP BY hour_of_day

 * mysql+pymysql://root:***@localhost:3306/md_water_services_two
14 rows affected.


[('09:00', Decimal('118')),
 ('10:00', Decimal('114')),
 ('11:00', Decimal('111')),
 ('12:00', Decimal('112')),
 ('13:00', Decimal('115')),
 ('14:00', Decimal('114')),
 ('15:00', Decimal('114')),
 ('16:00', Decimal('114')),
 ('07:00', Decimal('149')),
 ('18:00', Decimal('147')),
 ('08:00', Decimal('149')),
 ('17:00', Decimal('149')),
 ('06:00', Decimal('149')),
 ('19:00', Decimal('168'))]

# Water Accessibility and infrastructure summary report
#### This survey aimed to identify the water sources people use and determine both the total and average number of users for each source.

## Insights
#### 1. Most water sources are rural.
#### 2. 43% of our people are using shared taps. 2000 people often share one tap.
#### 3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes, pumps, and reservoirs.
#### 4. 18% of our people are using wells of which, but within that, only 28% are clean..
#### 5. Our citizens often face long wait times for water, averaging more than 120 minutes.
#### 6. In terms of queues:
#### - Queues are very long on Saturdays.
#### - Queues are longer in the mornings and evenings.
#### - Wednesdays and Sundays have the shortest queues. 